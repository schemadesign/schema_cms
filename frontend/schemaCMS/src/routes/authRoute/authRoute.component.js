import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Redirect, Route } from 'react-router-dom';
import { AUTH_PATH } from '../../shared/utils/api.constants';

export class ExternalRedirect extends PureComponent {
  static propTypes = {
    to: PropTypes.string.isRequired,
  };

  componentDidMount() {
    window.location.replace(this.props.to);
  }

  render = () => null;
}

export class AuthRoute extends PureComponent {
  static propTypes = {
    isAuthenticated: PropTypes.bool,
    requireAnonymous: PropTypes.bool,
    isUserFetched: PropTypes.bool,
    fetchCurrentUser: PropTypes.func.isRequired,
  };

  static defaultProps = {
    isAuthenticated: false,
    requireAnonymous: false,
    isUserFetched: false,
  };

  renderRoute = () => <Route {...this.props} />;

  render() {
    const { isUserFetched, requireAnonymous, isAuthenticated, fetchCurrentUser } = this.props;

    if (!isUserFetched) {
      fetchCurrentUser();
      return null;
    }

    if (isAuthenticated) {
      /**
       * This route is only viewable by unauthenticated users
       */
      if (requireAnonymous) {
        return <Redirect to="/" />;
      }

      /**
       * User has met all criteria and the route can be rendered
       */
      return this.renderRoute();
    }

    /**
     * This route is viewable by anonymous users
     */
    if (requireAnonymous) {
      return this.renderRoute();
    }

    /**
     * Redirect to register by default
     */
    return <ExternalRedirect to={AUTH_PATH} />;
  }
}
