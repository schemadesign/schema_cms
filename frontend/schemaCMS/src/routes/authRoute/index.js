export { default as AuthRoute } from './authRoute.container';
