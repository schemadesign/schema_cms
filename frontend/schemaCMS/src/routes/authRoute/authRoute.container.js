import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { createStructuredSelector } from 'reselect';
import { withRouter } from 'react-router-dom';
import { hot } from 'react-hot-loader';
import { compose } from 'ramda';

import { AuthRoute } from './authRoute.component';

import { selectIsAuthenticated } from '../../modules/userAuth';
import { selectIsFetched } from '../../modules/userProfile';
import { UserProfileRoutines } from '../../modules/userProfile/userProfile.redux';

const mapStateToProps = createStructuredSelector({
  isAuthenticated: selectIsAuthenticated,
  isUserFetched: selectIsFetched,
});

export const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      fetchCurrentUser: UserProfileRoutines.fetchUserDetails,
    },
    dispatch
  );

export default compose(hot(module), connect(mapStateToProps, mapDispatchToProps), withRouter)(AuthRoute);
