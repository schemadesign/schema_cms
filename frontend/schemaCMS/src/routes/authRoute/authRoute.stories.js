import React from 'react';
import { storiesOf } from '@storybook/react';

import { AuthRoute } from './authRoute.component';

const defaultProps = {};

storiesOf('AuthRoute', module).add('Default', () => <AuthRoute {...defaultProps} />);
