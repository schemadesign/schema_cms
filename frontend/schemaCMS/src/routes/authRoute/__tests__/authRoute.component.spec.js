import React from 'react';
import { shallow } from 'enzyme';
import { expect } from 'chai';
import { spy } from 'sinon';

import { AuthRoute } from '../authRoute.component';

describe('AuthRoute: Component', () => {
  const defaultProps = {
    isAuthenticated: false,
    requireAnonymous: false,
    isUserFetched: false,
    fetchCurrentUser: Function.prototype,
  };

  const component = props => <AuthRoute {...defaultProps} {...props} />;

  const render = (props = {}) => shallow(component(props));

  it('should render a <Redirect /> to home page when user is authenticated, fetched and require anonymous', () => {
    const wrapper = render({ isAuthenticated: true, isUserFetched: true, requireAnonymous: true });
    global.expect(wrapper).toMatchSnapshot();
  });

  it('should render null if user is not fetched', () => {
    const wrapper = render({ isUserFetched: false });
    global.expect(wrapper).toMatchSnapshot();
  });

  it('should call fetchCurrentUser if user is not fetched', () => {
    const fetchCurrentUser = spy();
    render({ isUserFetched: false, fetchCurrentUser });
    expect(fetchCurrentUser).to.have.been.calledOnce;
  });

  it('should render <Route> if user is logged and all criterias are met', () => {
    const wrapper = render({ isUserFetched: true, isAuthenticated: true, requireAnonymous: false });
    global.expect(wrapper).toMatchSnapshot();
  });

  it('should render <Route> if is anonymous user', () => {
    const wrapper = render({ requireAnonymous: true, isUserFetched: true });
    global.expect(wrapper).toMatchSnapshot();
  });

  it('should render <ExternalRedirect> if not authenticated', () => {
    const wrapper = render({ isUserFetched: true });
    global.expect(wrapper).toMatchSnapshot();
  });
});
