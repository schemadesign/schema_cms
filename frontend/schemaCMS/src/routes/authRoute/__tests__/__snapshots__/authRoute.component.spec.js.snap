// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthRoute: Component should render <ExternalRedirect> if not authenticated 1`] = `
<ExternalRedirect
  to="undefined/auth/login/auth0"
/>
`;

exports[`AuthRoute: Component should render <Route> if is anonymous user 1`] = `
<Route
  fetchCurrentUser={[Function]}
  isAuthenticated={false}
  isUserFetched={true}
  requireAnonymous={true}
/>
`;

exports[`AuthRoute: Component should render <Route> if user is logged and all criterias are met 1`] = `
<Route
  fetchCurrentUser={[Function]}
  isAuthenticated={true}
  isUserFetched={true}
  requireAnonymous={false}
/>
`;

exports[`AuthRoute: Component should render a <Redirect /> to home page when user is authenticated, fetched and require anonymous 1`] = `
<Redirect
  push={false}
  to="/"
/>
`;

exports[`AuthRoute: Component should render null if user is not fetched 1`] = `""`;
